### PR TITLE
Reimagined Activity Feed using memory access and without repeated data storage.

### PR DESCRIPTION
**Pull request by hunterriffic**
**UPDATE 7/15:** IT WORKS! Finally. All tests pass, and an Activity Feed is displayed. Exception handling is in place and the lists of activities are sorted before being passed through `getActivities`. Almost ready to submit.

*UPDATE 6/19:* The Activity Feed works (when tests are skipped)! One unit test- `testDoGetOneofEach`- still failing, even though there are no obvious differences between expected and actual. Need to add more unit tests for other cases once that test is passed. Also want to address the fact that the feed lists activities from earliest to latest and make it list latest activity first. There is also a spacing issue resulting from an earlier merge to master which needs to be addressed. Screenshot of Activity Feed below:
screen shot 2018-06-19 at 8 41 53 pm

*UPDATE 6/8:* Everything looks clean but I am getting a `NullPointerException` in my `testDoGet` in `ActivityFeedServlet`. See documentation below, any ideas as to why this is occurring would be appreciated.
screen shot 2018-06-09 at 12 21 31 am

I've made a bunch of file changes, primarily to `ActivityFeedServlet` and `activityfeed.jsp`. I added some unit tests. `ActivityFeedServletTest` is not passing `testDoGet`, and I continue to encounter NullPointerException errors when opening the Activity Feed for the first time, then NoSuchElementException once there has been activity in the app. There's a lot of work still left to be done, but I'd appreciate some feedback on glaring issues I may have missed.

### CODE REVIEW

File: src/main/java/codeu/model/store/basic/ConversationStore.java [hunk 1]
```diff
@@ -96,6 +97,16 @@ public Conversation getConversationWithTitle(String title) {
     return null;
   }
 
+  /** Find and return the Conversation with the given UUID. */
+  public Conversation getConversationWithId(UUID id) {
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/ConversationStore.java]: Please add a unit test for this method

File: src/test/java/codeu/model/data/ActivityTest.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package codeu.model.data;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ActivityTest {
```
Comment (willjschmitt) [inline src/test/java/codeu/model/data/ActivityTest.java]: Also test the getters

File: src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java [hunk 3]
```diff
@@ -103,4 +103,5 @@ public void writeThrough(Conversation conversation) {
   public void writeThrough(Message message) {
     persistentDataStore.writeThrough(message);
   }
+
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java]: Why is `PersistentStorageAgent` getting modified in this PR?

File: src/main/java/codeu/model/store/basic/UserStore.java [hunk 4]
```diff
@@ -32,7 +32,7 @@
 
   /**
    * Returns the singleton instance of UserStore that should be shared between all servlet classes.
-   * Do not call this function from a test; use getTestInstance() instead.
+   * Do not call this function fr om a test; use getTestInstance() instead.
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/UserStore.java]: This looks like an accident

File: src/main/java/codeu/model/store/basic/UserStore.java [hunk 5]
```diff
@@ -93,6 +93,21 @@ public User getUser(UUID id) {
     return null;
   }
 
+  /** Find and return User list. */
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/UserStore.java]: Please add unit tests for these new methods

File: src/main/java/codeu/model/store/basic/MessageStore.java [hunk 6]
```diff
@@ -85,6 +85,20 @@ public void addMessage(Message message) {
     return messagesInConversation;
   }
 
+  /** Find and return the Message with the given UUID. */
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/MessageStore.java]: Please add unit tests for these new methods

File: src/main/java/codeu/model/store/basic/MessageStore.java [hunk 7]
```diff
@@ -85,6 +85,20 @@ public void addMessage(Message message) {
...
+  public Message getMessageWithId(UUID id) {
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/MessageStore.java]: Is `getMessageWithId` used in this PR anywhere? Is it necessary?

Comment (willjschmitt) [reply src/main/java/codeu/model/store/basic/MessageStore.java]: If you think it will be useful for a future change, I'd ask you to split it into a separate PR. If you're not sure you'll need it, YAGNI principle should apply 😄

File: src/main/java/codeu/model/store/basic/UserStore.java [hunk 8]
```diff
@@ -93,6 +93,21 @@ public User getUser(UUID id) {
...
+  public List<User> getUsers() {
+    return users;
+  }
+
+  /** Find and return the User with the given UUID. */
+  public User getUserWithId(UUID id) {
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/UserStore.java]: Same comment about `getMessageWithId`

File: src/main/java/codeu/model/store/basic/UserStore.java [hunk 9]
```diff
@@ -93,6 +93,21 @@ public User getUser(UUID id) {
     return null;
   }
 
+  /** Find and return User list. */
+  public List<User> getUsers() {
```
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/UserStore.java]: `getAllUsers` to be consistent

File: src/main/java/codeu/model/store/basic/ConversationStore.java (hunk 1 above)
Comment (willjschmitt) [inline src/main/java/codeu/model/store/basic/ConversationStore.java]: Same comment as `getMessageWithId`

File: src/main/webapp/WEB-INF/view/activityfeed.jsp [hunk 10]
```diff
@@ -5,14 +5,23 @@
...
+
 <!DOCTYPE html>
 <html>
 <head>
   <title>Activity Feed</title>
   <link rel="stylesheet" href="/css/main.css">
 </head>
 <body>
-
+  <%@ include file="activity-helper.jsp" %>
```
Comment (willjschmitt) [inline src/main/webapp/WEB-INF/view/activityfeed.jsp]: I don't see this file

File: src/main/webapp/WEB-INF/view/activityfeed.jsp [hunk 11]
```diff
@@ -5,14 +5,23 @@
   Time: 7:51 PM
   To change this template use File | Settings | File Templates.
 --%>
+
+<%@ page import="codeu.model.data.Activity" %>
+<%@ page import="java.util.List" %>
+
+<%--
+Activity activity = (Activity) request.getAttribute("activity");
```
Comment (willjschmitt) [inline src/main/webapp/WEB-INF/view/activityfeed.jsp]: Should this commented code be here?

File: src/main/webapp/WEB-INF/view/activityfeed.jsp [hunk 12]
```diff
@@ -22,15 +31,45 @@
...
+          List<Activity> activities =
+            (List<Activity>) request.getAttribute("activities");
+          if(activities == null || activities.isEmpty()){
+          %>
+            <p>Get involved to see some activity.</p>
+          <%
+          }
+          else {
+          %>
+            Printing list<ul class="mdl-list">
```
Comment (willjschmitt) [inline src/main/webapp/WEB-INF/view/activityfeed.jsp]: You can probably drop "printing list"

File: src/main/java/codeu/model/data/Activity.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
+package codeu.model.data;
+
+/**
+ * Class representing an activity, which can be thought of as any action
+ * within the chat room. Activities are triggered by a User joining a
+ * conversation, starting a conversation or replying to a message.
+ */
+public class Activity {
```
Comment (willjschmitt) [inline src/main/java/codeu/model/data/Activity.java]: I would recommend creating a constructor that creates a fully initialized `Activity` rather than making them settable by consumers of it.

Then you can make `creationTime` and `output` `private final String`s

**Comment by willjschmitt** [comment]
Build is broken, also. PTAL

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 14]
```diff
@@ -1,16 +1,220 @@
...
+import java.util.LinkedList;
+import java.time.format.DateTimeFormatter;
+import java.time.ZoneId;
+import java.util.Objects;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+/*public class Activity {
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: Resolved

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 15]
```diff
@@ -1,16 +1,220 @@
...
+  public String creationTime;
+  public String output;
+}*/
+
 public class ActivityFeedServlet extends HttpServlet {
+
+    private enum Type {
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: Resolved

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 16]
```diff
@@ -1,16 +1,220 @@
...
+     * by the test framework or the servlet's init() function.
+     */
+    void setMessageStore(MessageStore messageStore) {
+        this.messageStore = messageStore;
+    }
+
+    /**
+     * Formats creation time into String for messages.
+     */
+    private String formatTimeMessage(Message message) {
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: In order to use one function for all 3 types I would pass in the UUID of the object and the `ActivityType`. If I passed in just `Instant creationTime` I would have to create functions  `getByCreationTime()` in each datastore. Trying the first method now.

[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: Resolved

File: src/main/java/codeu/controller/ChatServlet.java [hunk 17]
```diff
@@ -15,13 +15,13 @@
 package codeu.controller;
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ChatServlet.java]: That's a very good question.

File: src/main/java/codeu/model/store/basic/MessageStore.java (hunk 7 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/MessageStore.java]: `getMessageWithId` is never used, however `getConversationWithId` and `getUserWithId` both are.

File: src/main/java/codeu/model/store/basic/UserStore.java (hunk 8 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/UserStore.java]: Yes, it is used in `ActivityFeedServlet` in order to find the author of a message and owner of a conversation to include these details in the `output` string of an `Activity`.

File: src/main/java/codeu/model/store/basic/MessageStore.java (hunk 6 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/MessageStore.java]: The above method has been removed and will no longer require unit tests.

File: src/main/java/codeu/model/store/basic/ConversationStore.java (hunk 1 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/ConversationStore.java]: This function is used in `ActivityFeedServlet` in order to find which conversation a message in sent in to include this detail in the `output` string of an `Activity`.

File: src/main/java/codeu/controller/ChatServlet.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ChatServlet.java]: Made some spacing changes when first looking at the code. Should be reverted back to normal now.

File: src/main/java/codeu/controller/ConversationServlet.java [hunk 18]
```diff
@@ -14,6 +14,7 @@
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/ConversationServlet.java]: Reverted back to original.

File: src/main/java/codeu/controller/RegisterServlet.java [hunk 19]
```diff
@@ -9,6 +9,7 @@
 import javax.servlet.http.HttpServletRequest;
```
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/controller/RegisterServlet.java]: Reverted back to original.

File: src/main/java/codeu/model/data/Activity.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (hunterriffic) [reply src/main/java/codeu/model/data/Activity.java]: Agreed. It has been moved.

File: src/main/java/codeu/model/store/basic/UserStore.java (hunk 4 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/UserStore.java]: An accident it was indeed.

File: src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java (hunk 3 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/persistence/PersistentStorageAgent.java]: `PersistentStorageAgent` got modified when `ActivityStore` existed- after its removal some spacing mess had been left over. It has been reverted back to its original.

File: src/main/webapp/WEB-INF/view/activityfeed.jsp (hunk 12 above)
Comment (hunterriffic) [reply src/main/webapp/WEB-INF/view/activityfeed.jsp]: Dropped.

File: src/main/java/codeu/model/store/basic/ConversationStore.java (hunk 1 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/ConversationStore.java]: Unit test has now been added for this method.

File: src/main/java/codeu/model/store/basic/UserStore.java (hunk 5 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/UserStore.java]: Unit tests for these methods have been written.

File: src/main/java/codeu/model/store/basic/MessageStore.java (hunk 6 above)
Comment (hunterriffic) [reply src/main/java/codeu/model/store/basic/MessageStore.java]: Unit tests have been added.

File: src/main/webapp/WEB-INF/view/activityfeed.jsp (hunk 11 above)
Comment (hunterriffic) [reply src/main/webapp/WEB-INF/view/activityfeed.jsp]: Nope

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 20]
```diff
@@ -1,16 +1,203 @@
...
+    System.out.print("\n");
+    System.out.print("Length of messages: ");
+    System.out.print(messages.size());
+    System.out.print("\n");
+    System.out.print("Length of users: ");
+    System.out.print(users.size());
+    System.out.print("\n");
+
+    while (!conversations.isEmpty() || !users.isEmpty() || !messages.isEmpty()) {
+      // Iterates through all new conversations, users and messages in the lists, compares the creation times of
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: This describes the `while` block, so I would put this comment before it

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 21]
```diff
@@ -1,16 +1,203 @@
...
+      //  each then inserts objects chronologically into list of activities.
+      ActivityType nextActivity = null;
+      Instant nextActivityTime = null;
+
+      if (!conversations.isEmpty()) {
+        nextActivity = ActivityType.ConversationCreated;
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Why do the other blocks have null checks but not this one? I _might_ do it to just be  consistent

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 22]
```diff
@@ -1,16 +1,203 @@
...
+        nextActivityTime = conversations.peek().getCreationTime();
+      }
+      if (!users.isEmpty()) {
+        if (nextActivityTime == null) {
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Since the null check logic is the same check as the time check logic, I might combine them into the same conditional. Something like:

```
if (nextActivityType == null || users.peek().getCreationTime().compareTo(nextActivityTime) < 0) {
...
}
```

Comment (willjschmitt) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: Sane for the other block

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 23]
```diff
@@ -1,16 +1,203 @@
...
+          nextActivity = ActivityType.MessageSent;
+          nextActivityTime = messages.peek().getCreationTime();
+        }
+        if (messages.peek().getCreationTime().compareTo(nextActivityTime) < 0) {
+          nextActivity = ActivityType.MessageSent;
+          nextActivityTime = messages.peek().getCreationTime();
+        }
+      }
+
+      if (nextActivity != null) {
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Are there cases when this would be null? I expect not, and I might raise a fatal exception of some sort if it were, because that means something really bad happened

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 24]
```diff
@@ -1,16 +1,203 @@
...
+      }
     }
+    return activities;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException, ServletException {
+    List<Activity> activities = getActivities();
+    System.out.print(activities);
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Debug statement

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 25]
```diff
@@ -1,16 +1,203 @@
...
+    request.setAttribute("activities", activities);
+    request.getRequestDispatcher("/WEB-INF/view/activityfeed.jsp").forward(request, response);
+
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    System.out.print("in post request");
+    response.getOutputStream().println("in post request");
+    //response.sendRedirect("/activity");
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Remove commented line

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 26]
```diff
@@ -1,16 +1,203 @@
...
+    List<Activity> activities = getActivities();
+    System.out.print(activities);
+    request.setAttribute("activities", activities);
+    request.getRequestDispatcher("/WEB-INF/view/activityfeed.jsp").forward(request, response);
+
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    System.out.print("in post request");
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Remove debug staement

File: src/main/java/codeu/model/data/Activity.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+    return output;
+  }
+
+  public String toString() {
+    return creationTime + ": " + output;
+  }
+
+  @Override
+  public boolean equals(Object object) {
+    Activity activity = (Activity)object;
```
Comment (willjschmitt) [inline src/main/java/codeu/model/data/Activity.java]: What if the provided `Object` is not an `Activity`? Should we have a fatal exception of some sort?

File: src/test/java/codeu/controller/ActivityFeedServletTest.java [hunk 28]
```diff
@@ -1,34 +1,185 @@
...
+    String userOneContent = "test_username_one joined Charmer Chat.";
+    allActivities.add(new Activity(creationTimeTwo, userOneContent));
+    String conversationOneContent = "test_username_one started conversation conversation_one";
+    allActivities.add(new Activity(creationTimeThree, conversationOneContent));
+
+    activityFeedServlet.doGet(mockRequest, mockResponse);
+    Mockito.verify(mockRequest).setAttribute("activities", allActivities);
+  }
+
+  /*@After
```
Comment (willjschmitt) [inline src/test/java/codeu/controller/ActivityFeedServletTest.java]: This probably shouldn't be commented out. Why is it commented out? I would expect you need it

File: src/test/java/codeu/controller/ActivityFeedServletTest.java [hunk 29]
```diff
@@ -1,34 +1,185 @@
...
+    List<Activity> noActivities = new ArrayList<>();
+
+    activityFeedServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("activities", noActivities);
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoGetOneOfEach() throws IOException, ServletException {
```
Comment (willjschmitt) [inline src/test/java/codeu/controller/ActivityFeedServletTest.java]: Might be good to say why you aren't testing each activity type on its own (because you depend on the users, conversations, etc).

Or you should change these tests to use Mockito argument captors to check only the data you're interested in

File: src/test/java/codeu/controller/ActivityFeedServletTest.java [hunk 30]
```diff
@@ -1,34 +1,185 @@
...
+    activityFeedServlet.setMessageStore(mockMessageStore);
+
+    mockUserStore = UserStore.getTestInstance(mockPersistentStorageAgent);
+    activityFeedServlet.setUserStore(mockUserStore);
+
+    TimeZone.setDefault(chicagoTimeZone);
+  }
+
+  @Test
+  public void testDoGetNothing() throws IOException, ServletException {
```
Comment (willjschmitt) [inline src/test/java/codeu/controller/ActivityFeedServletTest.java]: Good test of a corner case

File: src/main/java/codeu/controller/ActivityFeedServlet.java [hunk 31]
```diff
@@ -1,16 +1,203 @@
...
+    String output = author.getName() + " sent in " + conversation.getTitle() + ": " + message.getContent();
+    return new Activity(creationTime,output);
+  }
+
+  /**
+   * Gets all activity on app, iterates through to sort by timestamp and prepares printed statements.
+   */
+  List<Activity> getActivities() {
+    List<Activity> activities = new LinkedList<>();
+    LinkedList<Conversation> conversations = new LinkedList<>(conversationStore.getAllConversations());
```
Comment (willjschmitt) [inline src/main/java/codeu/controller/ActivityFeedServlet.java]: Do we assume that these are sorted in some way? I don't see anything in the docstrings for the data stores that say they guarantee a returned order.

We might want to either guarantee that as part of the API (and add a test for it), or put sorts in place here

Comment (hunterriffic) [reply src/main/java/codeu/controller/ActivityFeedServlet.java]: They are now being sorted in the `GetActivities` function.
